### DISCUSSION TITLE
How destroy buffer in this classic use case ?

### DISCUSSION THREAD

**Discussion by ra1jin** [gpuweb/gpuweb#3243] [Q&A] [answered]
Hello, i need to destroy the vertex buffer in a debug draw function once his job is done.
I wanna known if any Promise based method exist to do that.
Actually my only solution is to make an array of buffers and destroy them at start of every frame.
Another solution is to create object for each debug drawable to hold his own vertex buffer accessible but for simple debug function it is a little bit overkill.

    this.passEncoder.setPipeline(this.lineDebugShaderRes.pipeline);

    let buf = new Float32Array(vertices);
    let vertexBuffer = this.device.createBuffer({ size: buf.byteLength, usage: GPUBufferUsage.VERTEX | GPUBufferUsage.COPY_DST });
    this.device.queue.writeBuffer(vertexBuffer, 0, buf.buffer);
    this.pendingBuffers.push(vertexBuffer);

    let mvpcMatrix = new Float32Array(Utils.MAT4_MULTIPLY(this.vpcMatrix, matrix));
    this.device.queue.writeBuffer(this.lineDebugShaderRes.uniformMVCPCMatrixBuffer, 0, mvpcMatrix.buffer);
    this.passEncoder.setBindGroup(0, this.lineDebugShaderRes.uniformMVCPCMatrixGroup);

    this.device.queue.writeBuffer(this.lineDebugShaderRes.uniformColorBuffer, 0, new Float32Array(color));
    this.passEncoder.setBindGroup(1, this.lineDebugShaderRes.uniformColorGroup);

    this.passEncoder.setVertexBuffer(0, vertexBuffer);
    this.passEncoder.draw(vertexCount);

    // vertexBuffer.destroy(); // if uncomment, webgpu make a warn (in a logical way)

**Comment by Kangz** [comment, marked as answer]
The buffers cannot be use in commands sent to a `GPUQueue.submit` when destroyed, but can be destroyed right after and the WebGPU implementation will make sure they live long enough. So the following code would work:

```js
let vertexBuffer = ....;

this.passEncoder.setVertexBuffer(0, vertexBuffer);
this.passEncoder.draw(vertexCount);
// More commands

// Valid since the buffer isn't destroyed yet.
this.device.queue.submit([this.passEncoder.finish()]);

vertexBuffer.destroy();
```
